vpdq/cpp/hashing/filehasher.cpp

```diff
@@ -62,322 +109,412 @@ static void saveFrameToFile(AVFrame* frame, const char* filename) {
+        } catch (const std::runtime_error& e) {
+          std::cerr << e.what() << std::endl;
+          throw;
+        }
+        // Resize the frame and convert to RGB24
+        sws_scale(
+            video->swsContext.get(),
+            decodeFrame->data,
+            decodeFrame->linesize,
+            0,
+            video->codecContext->height,
+            targetFrame->data,
+            targetFrame->linesize);
+
+        FatFrame fatFrame{std::move(targetFrame), get_frame_number()};
+        // Use the queue if multithreaded
+        if (multithread) {
+          std::lock_guard<std::mutex> lock(queue_mutex);
+          hash_queue.push(std::move(fatFrame));
+          queue_condition.notify_one();
+        } else {
+          hasher(std::move(fatFrame));
+        }
+      }
+    }
+
+    // This must be called to free the packet buffer filled by av_read_frame
+    av_packet_unref(packet.get());
+    return packet;
   }
 
-  // Create the output frame
-  AVFrame* frame = av_frame_alloc();
-  if (frame == nullptr) {
-    avcodec_free_context(&codecContext);
-    avformat_close_input(&formatContext);
-    return false;
+  void hasher(const FatFrame fatFrame) {
+    assert(fatFrame.frame->height != 0 && fatFrame.frame->width != 0);
+
+    std::unique_ptr<vpdq::hashing::AbstractFrameBufferHasher> phasher =
+        vpdq::hashing::FrameBufferHasherFactory::createFrameHasher(
+            fatFrame.frame->height, fatFrame.frame->width);
+
+    int quality;
+    pdq::hashing::Hash256 pdqHash;
+    bool ret = phasher->hashFrame(fatFrame.frame->data[0], pdqHash, quality);
+    if (!ret) {
+      throw std::runtime_error(
+          "Failed to hash frame buffer." + std::string("Frame: ") +
+          std::to_string(fatFrame.frameNumber) +
+          std::string(
+              " Frame width or height smaller than the minimum hashable dimension"));
+    }
+    if (verbose) {
+      std::cout << "PDQHash: " << pdqHash.format() << std::endl;
+    }
+
+    // Write frame to file here for debugging:
+    // This is not thread safe. Use one thread when writing to file
+    // saveFrameToFile(frame, "frame.rgb");
+
+    // Append vpdq feature to pdqHashes vector
+    std::lock_guard<std::mutex> lock(pdqHashes_mutex);
+    pdqHashes.emplace_back(vpdqFeature{
+        pdqHash,
+        static_cast<int>(fatFrame.frameNumber),
+        quality,
+        static_cast<double>(fatFrame.frameNumber) / video->frameRate});
   }
 
-  // Pixel format for the image passed to PDQ
-  constexpr AVPixelFormat pixelFormat = AV_PIX_FMT_RGB24;
+  void consumer() {
+    while (true) {
+      std::unique_lock<std::mutex> lock(queue_mutex);
+      queue_condition.wait(
+          lock, [this] { return !hash_queue.empty() || done_hashing; });
+      if (hash_queue.empty() && done_hashing)
+        break;
+      FatFrame fatFrame(std::move(hash_queue.front()));
+      hash_queue.pop();
+      lock.unlock();
```

Comment:
(unsure, alt): If you add one level of scoping you won't need the explicit unlock, i.e.

```
  { 
      std::unique_lock<std::mutex> lock(queue_mutex);
      queue_condition.wait(
          lock, [this] { return !hash_queue.empty() || done_hashing; });
      if (hash_queue.empty() && done_hashing)
        return;
      FatFrame fatFrame(std::move(hash_queue.front()));
      hash_queue.pop();
  }
  hasher(std::move(fatFrame));
```